ksqldb-parser/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java

```diff
@@ -129,15 +162,19 @@ private static Result coerceStruct(final Object value, final SqlStruct targetTyp
     return Result.of(coerced);
   }
 
-  private static Result coerceArray(final Object value, final SqlArray targetType) {
-    if (!(value instanceof List<?>)) {
+  private Result coerceArray(final Object value, final SqlArray targetType) {
+    final ListObject list;
+    if (value instanceof List<?>) {
+      list = JavaListObject.of((List<?>) value);
```

Comment:
I don't think it's necessary to create special classes JavaListObject/JsonListObject etc.

Can just call getList() on the JsonArray to get the List. This should simplify things quite a bit?

Reply:
I had this in an earlier version but it fails for structs nested inside arrays/maps because once getList() is called on the JsonArray, extracting a nested struct-type comes back as a LinkedHashMap rather than a JsonObject. When the LinkedHashMap is passed to coerceStruct(), coerceStruct() throws an exception because DefaultSqlValueCoercer is intended to fail if a caller attempts to coerce a map to a struct.

The other solutions I could think of for working around this were more complex than having the special classes JavaListObject/JsonListObject/etc.